### PR TITLE
fix(central): satisfaction survey link criteria

### PR DESCRIPTION
**Pull request by Rom1-B**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | !33982

The link to the search did not take into account “Duration of survey” defined on the entity.

image

image

### CODE REVIEW

File: src/Ticket.php [hunk 1]
```diff
@@ -4980,6 +4980,12 @@ public static function showCentralList($start, $status = "process", bool $showgr
                         $options['criteria'][1]['value']      = '^';
                         $options['criteria'][1]['link']       = 'AND';
 
+                        $duration = Entity::getUsedConfig('inquest_duration', $_SESSION['glpiactive_entity']);
+                        if ($duration > 0) {
+                            $options['criteria'][1]['searchtype'] = 'morethan';
+                            $options['criteria'][1]['value']      = '-' . $duration . 'DAY';
+                        }
```
Comment (cedric-anne) [inline src/Ticket.php:4987]: The problem is that this configuration may differ from one entity to another, and this link may display satisfactions from multiple entities at once. Depending on the entities configuration, the current proposal could filter surveys from a child entity that are still valid.
If we had the computed values for each entity, building a new search option to apply the correct filter would be easy, but we cannot change this in GLPI 10.0.
I do not have any other viable solution to propose.

For information, there is also a fixed for the list displayed in the central page in #17249 (not reviewed/validated yet).

**Review by cedric-anne — CHANGES_REQUESTED** [review]

**Review by AdrienClairembault — CHANGES_REQUESTED** [review]
Looks good but phpunit test cases should probably be added to validate the search criteria.

File: phpunit/functional/SearchTest.php
Comment (trasher) [inline phpunit/functional/SearchTest.php:1]: Since `Search` tests has not been migrated to phpunit yet; I'm not sure it is a good idea to do like this; it would be better to just add new case on existing (atoum) tests for now.

**Review by AdrienClairembault — APPROVED** [review]

File: phpunit/DbTestCase.php [hunk 2]
```diff
@@ -200,6 +200,7 @@ protected function createItem($itemtype, $input, $skip_fields = []): CommonDBTM
     {
         $item = new $itemtype();
         $id = $item->add($input);
+        $id = $item->getID();
```
Comment (AdrienClairembault) [inline phpunit/DbTestCase.php:203]: Is this change intentional ? seems weird

Comment (Rom1-B) [reply phpunit/DbTestCase.php:203]: Yes, because the add() function returns the value of the column named 'id', whereas getID() uses the column from getIndexName(), like getFromDB().
In most cases, getIndexName() returns 'id', but for the 'glpi_ticketsatisfactions' table, it's 'tickets_id'.

Comment (trasher) [reply phpunit/DbTestCase.php:203]: I guess we should keep the `id` field here. If you need the field from the `getId()` method, you'll have to retrieve it specifically.

Comment (Rom1-B) [reply phpunit/DbTestCase.php:203]: I did this so that the L121 checkInput() test would pass, since getFromDB() uses getIndexName()

Comment (Rom1-B) [reply phpunit/DbTestCase.php:203]: Unfortunately, getFromDByCrit() also uses getFromDB() 😢 

image

Comment (trasher) [reply phpunit/DbTestCase.php:203]: It uses `getFromDB()` with the `id` field - that sounds like a mistake and can lead to errors... All that stuff is always a nightmare; let's discuss about that when all core team will be present.

**Review by AdrienClairembault — APPROVED** [review]

File: phpunit/DbTestCase.php (hunk 2 above)
Comment (trasher) [reply phpunit/DbTestCase.php:203]: I've opened a PR to fix CommonDBTM, see https://github.com/glpi-project/glpi/pull/17696

File: phpunit/DbTestCase.php [hunk 3]
```diff
@@ -200,6 +200,7 @@ protected function createItem($itemtype, $input, $skip_fields = []): CommonDBTM
     {
         $item = new $itemtype();
         $id = $item->add($input);
```
Comment (cedric-anne) [inline phpunit/DbTestCase.php:202]:
```suggestion
        $item->add($input);
```

**Review by cedric-anne — DISMISSED** [review]
Seems OK. Could you rebase ?

**Review by cedric-anne — COMMENTED** [review]
Test validate the fix, so it seems OK to me. My SQL skills are too limited to fully understand the entire query.

**Review by AdrienClairembault — APPROVED** [review]
